File: apps/web/src/types/requisition.ts
export type RequisitionStatus = 'draft' | 'published' | 'closed';

export type EmploymentType = 'full_time' | 'part_time' | 'contract' | 'internship';

export interface Requisition {
  id: string;
  requisition_code: string;
  slug: string;
  title: string;
  department: string;
  location: string;
  employment_type: EmploymentType;
  experience_range: string;
  openings: number;
  hiring_manager: string;
  description_html: string;
  max_salary_budget?: number | null;
  hiring_complete_by?: string | null;
  status: RequisitionStatus;
  posted_at?: string | null;
  created_by?: string | null;
  created_at: string;
  updated_at: string;
  applications_count?: number;
}

export interface RequisitionFormData {
  title: string;
  department: string;
  location: string;
  employment_type: EmploymentType;
  experience_range: string;
  openings: number;
  hiring_manager: string;
  description_html: string;
  max_salary_budget?: number | null | string;
  hiring_complete_by?: string | null;
  status?: RequisitionStatus;
}

export interface RequisitionListResponse {
  items: Requisition[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export const DEPARTMENT_OPTIONS = [
  'Engineering',
  'Product',
  'Design',
  'Marketing',
  'Sales',
  'Operations',
  'Human Resources',
  'Finance',
  'Customer Support',
  'Legal',
  'Data & Analytics',
  'Security',
] as const;

export const EMPLOYMENT_TYPE_LABELS: Record<EmploymentType, string> = {
  full_time: 'Full-time',
  part_time: 'Part-time',
  contract: 'Contract',
  internship: 'Internship',
};

export const EXPERIENCE_RANGE_OPTIONS = [
  '0-1 years (Entry level)',
  '1-3 years (Junior)',
  '3-5 years (Mid-level)',
  '5-8 years (Senior)',
  '8-12 years (Staff / Principal)',
  '12+ years (Executive / Director)',
] as const;

File: apps/api/app/schemas/requisition.py
import uuid
from datetime import date, datetime
from decimal import Decimal
from typing import Optional
from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import EmploymentType, RequisitionStatus


class RequisitionBase(BaseModel):
    title: str = Field(..., min_length=1, max_length=100)
    department: str = Field(..., min_length=1, max_length=80)
    location: str = Field(..., min_length=1, max_length=120)
    employment_type: EmploymentType
    experience_range: str = Field(..., min_length=1, max_length=40)
    openings: int = Field(..., gt=0)
    hiring_manager: str = Field(..., min_length=1, max_length=120)
    description_html: str = Field(...)
    max_salary_budget: Optional[Decimal] = None
    hiring_complete_by: Optional[date] = None


class RequisitionCreate(RequisitionBase):
    status: Optional[RequisitionStatus] = RequisitionStatus.DRAFT


class RequisitionUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=100)
    department: Optional[str] = Field(None, min_length=1, max_length=80)
    location: Optional[str] = Field(None, min_length=1, max_length=120)
    employment_type: Optional[EmploymentType] = None
    experience_range: Optional[str] = Field(None, min_length=1, max_length=40)
    openings: Optional[int] = Field(None, gt=0)
    hiring_manager: Optional[str] = Field(None, min_length=1, max_length=120)
    description_html: Optional[str] = None
    max_salary_budget: Optional[Decimal] = None
    hiring_complete_by: Optional[date] = None
    status: Optional[RequisitionStatus] = None


class RequisitionResponse(RequisitionBase):
    id: uuid.UUID
    requisition_code: str
    slug: str
    status: RequisitionStatus
    posted_at: Optional[datetime] = None
    created_by: Optional[uuid.UUID] = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class RequisitionAdminResponse(RequisitionResponse):
    application_count: int = 0

    model_config = ConfigDict(from_attributes=True)


class PublicJobListItem(BaseModel):
    id: uuid.UUID
    requisition_code: str
    slug: str
    title: str
    department: str
    location: str
    employment_type: EmploymentType
    experience_range: str
    openings: int
    posted_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class PublicJobDetail(BaseModel):
    id: uuid.UUID
    requisition_code: str
    slug: str
    title: str
    department: str
    location: str
    employment_type: EmploymentType
    experience_range: str
    openings: int
    description_html: str
    posted_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class RequisitionFilter(BaseModel):
    status: Optional[RequisitionStatus] = None
    department: Optional[str] = None
    q: Optional[str] = None
    skip: int = 0
    limit: int = 100
